### PR TITLE
Alerting: Enable alerts preview on notification policies page

### LINKED ISSUE #68289: Enable alerts preview on the notifications policy page
https://github.com/grafana/grafana/pull/61952 added alerts preview on the notifications policy page hidden behind a feature toggle due to performance issues.
We want to solve the issues and enable the preview without the feature toggle

### PR DESCRIPTION
**Pull request by konrad147**
**What is this feature?**
This PR enables alerts preview on the notifications policy page. Previously the feature was hidden behind a feature toggle due to performance issues on large notification route trees.
This PR offloads the matching calculation to a web worker to keep the UI smooth during the alerts matching.

**Which issue(s) does this PR fix?**:
Fixes #68289

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/alerting/unified/components/notification-policies/Policy.tsx [hunk 1]
```diff
@@ -43,7 +43,7 @@ interface PolicyComponentProps {
   contactPointsState?: ReceiversState;
   readOnly?: boolean;
   inheritedProperties?: InhertitableProperties;
-  routesMatchingFilters?: RouteWithID[];
+  routesMatchingFilters?: string[];
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/components/notification-policies/Policy.tsx]: why we don't need this recursive type anymore?

File: public/app/features/alerting/unified/components/notification-policies/Policy.tsx [hunk 2]
```diff
@@ -114,9 +116,10 @@ const Policy: FC<PolicyComponentProps> = ({
   const isEditable = canEditRoutes;
   const isDeletable = canDeleteRoutes && !isDefaultPolicy;
 
-  const matchingAlertGroups = useMemo(() => {
-    return showMatchingInstances ? findMatchingAlertGroups(routeTree, currentRoute, alertGroups) : [];
-  }, [alertGroups, currentRoute, routeTree, showMatchingInstances]);
+  // const matchingAlertGroups = useMemo(() => {
+  //   return showMatchingInstances ? findMatchingAlertGroups(routeTree, currentRoute, alertGroups) : [];
+  // }, [alertGroups, currentRoute, routeTree, showMatchingInstances]);
+  const matchingAlertGroups = routeAlertGroupsMap.get(currentRoute.id) ?? [];
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/components/notification-policies/Policy.tsx]: 🙌

File: public/app/features/alerting/unified/components/notification-policies/Policy.tsx [hunk 3]
```diff
@@ -114,12 +113,15 @@ const Policy: FC<PolicyComponentProps> = ({
   const isEditable = canEditRoutes;
   const isDeletable = canDeleteRoutes && !isDefaultPolicy;
 
-  const matchingAlertGroups = useMemo(() => {
-    return showMatchingInstances ? findMatchingAlertGroups(routeTree, currentRoute, alertGroups) : [];
-  }, [alertGroups, currentRoute, routeTree, showMatchingInstances]);
+  // const matchingAlertGroups = useMemo(() => {
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/notification-policies/Policy.tsx:116]: TODO Remove

File: public/app/features/alerting/unified/components/notification-policies/Policy.tsx [hunk 4]
```diff
@@ -64,14 +62,15 @@ const Policy: FC<PolicyComponentProps> = ({
...
+  routeAlertGroupsMap,
   onEditPolicy,
   onAddPolicy,
   onDeletePolicy,
   onShowAlertInstances,
 }) => {
   const styles = useStyles2(getStyles);
   const isDefaultPolicy = currentRoute === routeTree;
-  const showMatchingInstances = useEnabled(AlertingFeature.NotificationPoliciesV2MatchingInstances);
+  const showMatchingInstances = true; // useEnabled(AlertingFeature.NotificationPoliciesV2MatchingInstances);
```
Comment (konrad147) [inline public/app/features/alerting/unified/components/notification-policies/Policy.tsx:73]: TODO Remove

File: public/app/features/alerting/unified/components/notification-policies/Policy.tsx (hunk 1 above)
Comment (konrad147) [reply public/app/features/alerting/unified/components/notification-policies/Policy.tsx]: I think this is outdated version. In the latest one there is no change in this line

File: public/app/features/alerting/unified/utils/notification-policies.ts [hunk 5]
```diff
@@ -72,6 +36,24 @@ function findMatchingRoutes<T extends Route>(root: T, labels: Label[]): T[] {
   return matches;
 }
 
+//export type NormalizedRoute = Omit<RouteWithID, 'matchers' | 'match' | 'match_re'> & { routes?: NormalizedRoute[] };
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/utils/notification-policies.ts:39]: comment

File: public/app/features/alerting/unified/routeGroupsMatcherWorker.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import * as comlink from 'comlink';
+
+import type { AlertmanagerGroup, RouteWithID } from '../../../plugins/datasource/alertmanager/types';
+
+import { findMatchingAlertGroups, normalizeRootRoute } from './utils/notification-policies';
+
+const routeGroupsMatcher = {
+  getRouteGroupsMap(rootRoute: RouteWithID, groups: AlertmanagerGroup[]): Map<string, AlertmanagerGroup[]> {
+    const normalizedRootRoute = normalizeRootRoute(rootRoute);
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/routeGroupsMatcherWorker.ts:9]: why we only normalize the root route?

**Review by soniaAguilarPeiron — APPROVED** [review]
🙌  LGTM! 🚀

**Review by gillesdemey — APPROVED** [review]
Code changes LGTM; nice bit of cleanup and refactoring too ✨

**Pull request merged**